### Hyperparameter:

- They are the parameter specified inside the ML Algorithm, now since these parameter affect the accuracy of the ML algorithm hence we call them as Hyperparameters.
- Tuning : Adjusting

#### Hyperparameter Tuning : To adjust the parameter of any ML Algorithm in such a way that the specified parameter give us the max accuracy.

### Load the standard libraries

In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
warnings.filterwarnings('ignore')

## Load the data

In [3]:
data = pd.read_csv('Health_Insurance.csv')
data.head()

Age  Purchased
0   19          0
1   35          0
2   26          0
3   27          0
4   19          0

## Seperate X and y

In [4]:
X = data.drop('Purchased', axis = 1)
y = data['Purchased']

### Split the data into train and test set

In [5]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 0)

## Apply KNeighborsClassifier Algorithm on the data

In [7]:
from sklearn.neighbors import KNeighborsClassifier
knc = KNeighborsClassifier()   ### default parameter values are 5 neighbors and Euclidean distance
knc

KNeighborsClassifier()

In [8]:
knc.fit(X_train, y_train)

KNeighborsClassifier()

## Perform predictions on X_test

In [9]:
y_pred = knc.predict(X_test)
y_pred

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 1, 0, 1, 1, 0, 1, 0, 0, 1,
       0, 1, 0, 1, 0, 0, 1, 1, 0, 1, 1, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0,
       1, 0, 1, 1, 0, 1, 1, 0, 0, 0, 1, 1, 0, 0, 1, 0, 1, 1, 0, 0, 0, 1,
       0, 0, 0, 1, 1, 0, 0, 1, 0, 0, 1, 0, 1, 1], dtype=int64)

### Perform Evaluations

In [10]:
from sklearn.metrics import accuracy_score
accuracy_score(y_test, y_pred)

0.825

#### Note:
Since, KNeighborsClassifier is applied on the data, the accuracy we get is 82.5% when the n_neighbors in KNC is 5 and distance metric used is Euclidean distance.

# Hyperparameter Tuning:

- To figure out excatly at what value of n_neigbors and distance metric does the KNeighborsClassifier Algorithm perform better?

In [18]:
from sklearn.neighbors import KNeighborsClassifier
knc = KNeighborsClassifier()

### Hyperparameter optimization
parameters = {'n_neighbors' : [3, 5, 7, 9, 11, 13, 15, 17, 19, 21]}

### To apply the KNeighborsClassifier on the data with different value of n_neighbors we use the a function called GridSearchCV

In [19]:
from sklearn.model_selection import GridSearchCV
gscv = GridSearchCV(knc, parameters)
gscv

GridSearchCV(estimator=KNeighborsClassifier(),
             param_grid={'n_neighbors': [3, 5, 7, 9, 11, 13, 15, 17, 19, 21]})

In [20]:
gscv.fit(X_train, y_train)

GridSearchCV(estimator=KNeighborsClassifier(),
             param_grid={'n_neighbors': [3, 5, 7, 9, 11, 13, 15, 17, 19, 21]})

In [21]:
gscv.cv_results_

{'mean_fit_time': array([0.00315514, 0.00339098, 0.00328689, 0.00227342, 0.00238938,
        0.00208535, 0.00217776, 0.00209789, 0.00205932, 0.00245786]),
 'std_fit_time': array([0.00084826, 0.00094638, 0.00139289, 0.00049971, 0.00063912,
        0.00032821, 0.00019087, 0.00044403, 0.00055071, 0.00022895]),
 'mean_score_time': array([0.00648646, 0.00670638, 0.00502601, 0.00465798, 0.00456047,
        0.00448565, 0.00467649, 0.00485244, 0.00454884, 0.00446653]),
 'std_score_time': array([0.00137384, 0.00194152, 0.00108306, 0.00065522, 0.00032069,
        0.00021238, 0.00055169, 0.00035615, 0.00090963, 0.00028569]),
 'param_n_neighbors': masked_array(data=[3, 5, 7, 9, 11, 13, 15, 17, 19, 21],
              mask=[False, False, False, False, False, False, False, False,
                    False, False],
        fill_value='?',
             dtype=object),
 'params': [{'n_neighbors': 3},
  {'n_neighbors': 5},
  {'n_neighbors': 7},
  {'n_neighbors': 9},
  {'n_neighbors': 11},
  {'n_neighbors'

In [22]:
res = pd.DataFrame(gscv.cv_results_)
res[['params', 'mean_test_score', 'rank_test_score']]

params  mean_test_score  rank_test_score
0   {'n_neighbors': 3}         0.753125               10
1   {'n_neighbors': 5}         0.803125                7
2   {'n_neighbors': 7}         0.800000                8
3   {'n_neighbors': 9}         0.787500                9
4  {'n_neighbors': 11}         0.806250                6
5  {'n_neighbors': 13}         0.818750                1
6  {'n_neighbors': 15}         0.815625                2
7  {'n_neighbors': 17}         0.815625                2
8  {'n_neighbors': 19}         0.815625                2
9  {'n_neighbors': 21}         0.815625                2

#### At k = 13 the KNN algorithm is performing better on the data. Hence the optimized value of hyperparameter k is 13.

- When the k value is 15 and above the accuracy stops increasing or is constant at k = 15, k = 17, k = 19 etc. This indicates that k = 15 and below is the one where the accuracy is changing

When the k = 13, for k = 13 weather Euclidean distance perform better or Manhattan distance perform better

In [24]:
from sklearn.neighbors import KNeighborsClassifier
knc = KNeighborsClassifier()

## Hyperparameter optimization
parameters = {'n_neighbors' : [3, 5, 7, 9, 11, 13, 15, 17], 'p' : [1, 2]}

In [25]:
from sklearn.model_selection import GridSearchCV
gscv = GridSearchCV(knc, parameters)
gscv

GridSearchCV(estimator=KNeighborsClassifier(),
             param_grid={'n_neighbors': [3, 5, 7, 9, 11, 13, 15, 17],
                         'p': [1, 2]})

In [26]:
gscv.fit(X_train, y_train)

GridSearchCV(estimator=KNeighborsClassifier(),
             param_grid={'n_neighbors': [3, 5, 7, 9, 11, 13, 15, 17],
                         'p': [1, 2]})

In [27]:
gscv.cv_results_

{'mean_fit_time': array([0.00370474, 0.00257301, 0.00203681, 0.00261841, 0.00267763,
        0.00252438, 0.00204997, 0.00202341, 0.00211391, 0.00263782,
        0.00337543, 0.00245519, 0.00263124, 0.00417533, 0.00284967,
        0.00286779]),
 'std_fit_time': array([0.00084462, 0.00100119, 0.00039343, 0.00032846, 0.00061057,
        0.00026252, 0.00033765, 0.00071853, 0.00019967, 0.00039414,
        0.00085914, 0.00057716, 0.00088631, 0.00285101, 0.00096726,
        0.00116578]),
 'mean_score_time': array([0.00608783, 0.00479355, 0.00476604, 0.00518718, 0.004598  ,
        0.00415196, 0.00424457, 0.00446463, 0.00509362, 0.00534534,
        0.00622549, 0.00548205, 0.00613298, 0.00710301, 0.00629783,
        0.00524316]),
 'std_score_time': array([0.00126196, 0.00084194, 0.00042638, 0.00182412, 0.00074674,
        0.00020359, 0.00069012, 0.00075283, 0.00128516, 0.00056707,
        0.00115838, 0.00101302, 0.00121056, 0.00298787, 0.00116844,
        0.00054686]),
 'param_n_neighbors': mask

In [28]:
res = pd.DataFrame(gscv.cv_results_)
res

mean_fit_time  std_fit_time  mean_score_time  std_score_time  \
0        0.003705      0.000845         0.006088        0.001262   
1        0.002573      0.001001         0.004794        0.000842   
2        0.002037      0.000393         0.004766        0.000426   
3        0.002618      0.000328         0.005187        0.001824   
4        0.002678      0.000611         0.004598        0.000747   
5        0.002524      0.000263         0.004152        0.000204   
6        0.002050      0.000338         0.004245        0.000690   
7        0.002023      0.000719         0.004465        0.000753   
8        0.002114      0.000200         0.005094        0.001285   
9        0.002638      0.000394         0.005345        0.000567   
10       0.003375      0.000859         0.006225        0.001158   
11       0.002455      0.000577         0.005482        0.001013   
12       0.002631      0.000886         0.006133        0.001211   
13       0.004175      0.002851         0.007103        0.002988   
14       0.002850      0.000967         0.006298        0.001168   
15       0.002868      0.001166         0.005243        0.000547   

   param_n_neighbors param_p                       params  split0_test_score  \
0                  3       1   {'n_neighbors': 3, 'p': 1}           0.765625   
1                  3       2   {'n_neighbors': 3, 'p': 2}           0.765625   
2                  5       1   {'n_neighbors': 5, 'p': 1}           0.765625   
3                  5       2   {'n_neighbors': 5, 'p': 2}           0.765625   
4                  7       1   {'n_neighbors': 7, 'p': 1}           0.781250   
5                  7       2   {'n_neighbors': 7, 'p': 2}           0.781250   
6                  9       1   {'n_neighbors': 9, 'p': 1}           0.765625   
7                  9       2   {'n_neighbors': 9, 'p': 2}           0.765625   
8                 11       1  {'n_neighbors': 11, 'p': 1}           0.781250   
9                 11       2  {'n_neighbors': 11, 'p': 2}           0.781250   
10                13       1  {'n_neighbors': 13, 'p': 1}           0.781250   
11                13       2  {'n_neighbors': 13, 'p': 2}           0.781250   
12                15       1  {'n_neighbors': 15, 'p': 1}           0.765625   
13                15       2  {'n_neighbors': 15, 'p': 2}           0.765625   
14                17       1  {'n_neighbors': 17, 'p': 1}           0.765625   
15                17       2  {'n_neighbors': 17, 'p': 2}           0.765625   

    split1_test_score  split2_test_score  split3_test_score  \
0            0.718750           0.687500           0.781250   
1            0.718750           0.687500           0.781250   
2            0.750000           0.812500           0.796875   
3            0.750000           0.812500           0.796875   
4            0.750000           0.765625           0.812500   
5            0.750000           0.765625           0.812500   
6            0.765625           0.718750           0.796875   
7            0.765625           0.718750           0.796875   
8            0.765625           0.781250           0.796875   
9            0.765625           0.781250           0.796875   
10           0.765625           0.828125           0.812500   
11           0.765625           0.828125           0.812500   
12           0.765625           0.828125           0.812500   
13           0.765625           0.828125           0.812500   
14           0.765625           0.828125           0.812500   
15           0.765625           0.828125           0.812500   

    split4_test_score  mean_test_score  std_test_score  rank_test_score  
0            0.812500         0.753125        0.044634               15  
1            0.812500         0.753125        0.044634               15  
2            0.890625         0.803125        0.049014                9  
3            0.890625         0.803125        0.049014                9  
4            0.890625         0.800000      

In [29]:
res[['params', 'mean_test_score', 'rank_test_score']]

params  mean_test_score  rank_test_score
0    {'n_neighbors': 3, 'p': 1}         0.753125               15
1    {'n_neighbors': 3, 'p': 2}         0.753125               15
2    {'n_neighbors': 5, 'p': 1}         0.803125                9
3    {'n_neighbors': 5, 'p': 2}         0.803125                9
4    {'n_neighbors': 7, 'p': 1}         0.800000               11
5    {'n_neighbors': 7, 'p': 2}         0.800000               11
6    {'n_neighbors': 9, 'p': 1}         0.787500               13
7    {'n_neighbors': 9, 'p': 2}         0.787500               13
8   {'n_neighbors': 11, 'p': 1}         0.806250                7
9   {'n_neighbors': 11, 'p': 2}         0.806250                7
10  {'n_neighbors': 13, 'p': 1}         0.818750                1
11  {'n_neighbors': 13, 'p': 2}         0.818750                1
12  {'n_neighbors': 15, 'p': 1}         0.815625                3
13  {'n_neighbors': 15, 'p': 2}         0.815625                3
14  {'n_neighbors': 17, 'p': 1}         0.815625                3
15  {'n_neighbors': 17, 'p': 2}         0.815625                3

### Note:

- Changing the distance from Manhattan to Euclidean distance, the accuracy remains constant indicating
that accuracy is not dependent on the distance metric.
- The accuracy is highest when k = 13 and distance is either Manhattan or Euclidean (any one can be taken)